### PR TITLE
Replace Jython with Jep to enable use of modern python in checks 

### PR DESCRIPTION
**Pull request by jeanetteclark**
yay

### CODE REVIEW

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java [hunk 1]
```diff
@@ -1,57 +1,77 @@
...
-     * 2) function definition for call() that uses given global variables
-     * 3) fully-qualified Java bean name that implements Callable<Result> and has properties that 
-     * match the given variable names
-     * @param variables the variable name/values that will be made available to the script
-     * @param code the code, function definition, or classname
-     * @return
-     * @throws ScriptException
-     */
+	// protected ScriptEngineManager manager = new ScriptEngineManager();
+	protected JepScriptEngineFactory jepSEF;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:39]: Field 'jepSEF' is never used

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java [hunk 2]
```diff
@@ -1,57 +1,77 @@
...
+	 * 1) script code that uses given global variables
+	 * 2) function definition for call() that uses given global variables
+	 * 3) fully-qualified Java bean name that implements Callable<Result> and has
+	 * properties that
+	 * match the given variable names
+	 * 
+	 * @param variables the variable name/values that will be made available to the
+	 *                  script
+	 * @param code      the code, function definition, or classname
+	 * @return
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:56]: '@return' tag description is missing

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java [hunk 3]
```diff
@@ -79,85 +102,136 @@ public Result dispatch(Map<String, Object> variables, String code) throws Script
...
 		String currentEnv = engine.getFactory().getLanguageName();
 		log.debug("currentEnv=" + currentEnv);
 		return env.equalsIgnoreCase(currentEnv);
 	}
-	
-	protected Dispatcher() {}
 
+	public void close() {
+		if (engine instanceof JepScriptEngine) {
+			JepScriptEngine jepEngine = (JepScriptEngine) engine;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:188]: Today I learned from my IDE..

lines 187 & 188 can be replaced simply by:

```java
		if (engine instanceof JepScriptEngine jepEngine) {
```

How cool is that?? 😁

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:1]: global search & replace tabs -> 4 spaces

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java [hunk 4]
```diff
@@ -197,4 +272,36 @@ public Map<String, Object> getBindings() {
 	public void setBindings(Map<String, Object> bindings) {
 		this.bindings = bindings;
 	}
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:274]: Method `setBindings(java.util.Map<java.lang.String,java.lang.Object>)` is never used

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java [hunk 5]
```diff
@@ -197,4 +272,36 @@ public Map<String, Object> getBindings() {
...
+
+		if (!Files.exists(Path.of(jepPath))) {
+			jepPath = pythonFolder + "/libjep.so";
+		}
+
+		// set path for jep executing python
+		try {
+			MainInterpreter.setJepLibraryPath(jepPath);
+		} catch (JepException e) {
+			throw new MetadigException("Error confguring Jep: " + e);
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:303]: typo `confguring`

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/JepScriptEngine.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            this.factory = new JepScriptEngineFactory();
+        return this.factory;
+    }
+
+    /**
+     * Get bindings
+     * 
+     * Not sure exactly how/if this works, got this from an old Jep release
+     * 
+     * @param scope
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/JepScriptEngine.java:148]: missing description

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/JepScriptEngine.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    @Override
+    public Bindings getBindings(int scope) {
+        if (scope == ScriptContext.ENGINE_SCOPE) {
+            return this.bindings;
+        }
+
+        return this.globalBindings;
+    }
+
+    public String getEngineName() {
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/dispatch/JepScriptEngine.java:160]: Method 'getEngineName()' is never used

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:1]: tabs to spaces

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 8]
```diff
@@ -51,84 +55,87 @@ public Result call() {
...
+					// Look up cross-reference info as well
 					outputs.addAll(this.lookupCrossRef(id));
 
-				} catch (Exception e) {
-					log.error("Could not look up award " + awardId + " at URL " + url.toString() + ": ", e);
+				} catch (MalformedURLException me) {
+					log.error("Could not look up award " + awardId + " at URL " + url.toString()
+							+ "malformed URL exception");
+					log.error(me.getMessage());
+					continue;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:111]: `continue` is unnecessary as the last statement in a loop

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 9]
```diff
@@ -51,84 +55,87 @@ public Result call() {
...
+				} catch (IOException ioe) {
+					log.error("Could not look up award " + awardId + " at URL " + url.toString() + "IO exception");
+					log.error(ioe.getMessage());
+					continue;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:115]: `continue` is unnecessary as the last statement in a loop

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 10]
```diff
@@ -51,84 +55,87 @@ public Result call() {
...
+				} catch (ParseException pe) {
+					log.error("Could not look up award " + awardId + " at URL " + url.toString() + "parse exception");
+					log.error(pe.getMessage());
 					continue;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:119]: `continue` is unnecessary as the last statement in a loop

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 11]
```diff
@@ -51,84 +55,87 @@ public Result call() {
...
-		
+
 		return result;
-		
+
 	}
-	
+
 	// TODO: implement this in a way that makes sense
 	private List<Output> lookupCrossRef(String awardId) {
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:137]: This method does a bunch of string manipulation, bet does not actually add any of those strings to the return value. It simply returns the empty ArrayList `outputs`.

If it's not needed, I'd recommend removing it completely (it's only called from 1 place)

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 12]
```diff
@@ -152,5 +159,5 @@ public Object getAwards() {
 	public void setAwards(Object awards) {
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:159]: setAwards (and getAwards, above) are never used. Can they be deleted?

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java [hunk 13]
```diff
@@ -137,8 +144,8 @@ private List<Output> lookupCrossRef(String awardId) {
 
 		// message.items[0].funder[x].DOI
 		String funderDOI = null;
-		String funderId = null; //funderDOI.substring(funderDOI.lastIndexOf("/"));
-		
+		String funderId = null; // funderDOI.substring(funderDOI.lastIndexOf("/"));
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/AwardLookupCheck.java:147]: never used

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java [hunk 14]
```diff
@@ -38,63 +38,64 @@
...
 	}
-	
+
 	private void extractNamespaces() {
 		XPath xpath = xPathfactory.newXPath();
 		NodeList nodes = null;
 		try {
-			//String selectorPath = "//*[namespace-uri()]/concat(substring-before(name(), ':'),':',namespace-uri())";
+			// String selectorPath = "//*[namespace-uri()]/concat(substring-before(name(),
+			// ':'),':',namespace-uri())";
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:89]: delete commented code

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:1]: global search & replace tabs to spaces

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java [hunk 15]
```diff
@@ -370,9 +382,10 @@ private Object selectPath(Selector selector, Node contextNode) throws XPathExpre
 				Namespace entry = nsIter.next();
 				String prefix = entry.getPrefix();
 				String uri = entry.getUri();
-                // Some metadata files may have improper xmlns declarations that don't include
+				// Some metadata files may have improper xmlns declarations that don't include
 				// a prefix (encountered in Dryad Data), so skip these.
-				if(prefix == null) continue;
+				if (prefix == null)
+					continue;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:388]: missing brackets

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java [hunk 16]
```diff
@@ -420,42 +432,52 @@ else if (nodes.getLength() > 0 || selector.getSubSelector() != null) {
...
-		if (NumberUtils.isNumber((String)value)) {
-			result = NumberUtils.createNumber((String)value);
-		} else {
-			// relies on this method to return null if we are not sure if it is a boolean
-			Boolean bool = BooleanUtils.toBooleanObject((String)value);
-			if (bool != null) {
-				result = bool;
+
+		if (value instanceof String) {
+			String stringValue = (String) value;
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:465]: Variable 'stringValue' can be replaced with pattern variable
```java
		if (value instanceof String stringValue) {
```

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java [hunk 17]
```diff
@@ -481,14 +503,14 @@ public void setSystemMetadata(SystemMetadata systemMetadata) {
...
-		} catch(TransformerException ex) {
-		    ex.printStackTrace();
-		    return null;
+			Transformer transformer = TransformerFactory.newInstance().newTransformer();
+			StreamResult result = new StreamResult(new StringWriter());
+			DOMSource source = new DOMSource(document);
+			transformer.transform(source, result);
+			return result.getWriter().toString();
+		} catch (TransformerException ex) {
+			ex.printStackTrace();
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:512]: can this be replaced with a logging call?

**Review by artntek — CHANGES_REQUESTED** [review]

File: src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java (hunk 3 above)
Comment (jeanetteclark) [reply src/main/java/edu/ucsb/nceas/mdqengine/dispatch/Dispatcher.java:188]: ooh nice!

File: src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java [hunk 18]
```diff
@@ -509,7 +507,7 @@ private String toXmlString(Document document) {
 			transformer.transform(source, result);
 			return result.getWriter().toString();
 		} catch (TransformerException ex) {
-			ex.printStackTrace();
+			log.error("Error transforming to XML string." + ex.getMessage());
```
Comment (artntek) [inline src/main/java/edu/ucsb/nceas/mdqengine/processor/XMLDialect.java:510]: needs a space after `string.`

**Review by artntek — APPROVED** [review]
one small typo to change. I approved

**Pull request merged**